src/material/input/input.ts

```diff
@@ -83,7 +83,7 @@ const _MatInputBase = mixinErrorState(class {
     '[attr.readonly]': 'readonly && !_isNativeSelect || null',
     // Only mark the input as invalid for assistive technology if it has a value since the
     // state usually overlaps with `aria-required` when the input is empty and can be redundant.
-    '[attr.aria-invalid]': 'errorState && !empty',
+    '[attr.aria-invalid]': 'empty ? null : errorState',
```

Comment:
While we're here, is it _slightly_ more correct to change this to `'(empty && required) ? null : errorState'`? I'm thinking of an edge case where an input is empty, but not required, but still in an error state (maybe because there's validation that looks for duplicates and there's already an empty input). If so, we'd want to update the comment above as well.

Reply:
This behavior is intentional, see https://github.com/angular/components/pull/21609. There's also a linked issue.

Reply:
Yep!  The reason #21609 was requested was because without it `aria-invalid` was being applied to `aria-required` inputs with no value.  This violates the ARIA spec (see here : https://www.w3.org/TR/wai-aria/#aria-invalid ) and caused #18140 to be filed.

Unfortunately, setting `aria-invalid="false"` on these inputs suppresses any native indication to a screen reader user that the input is in an invalid state.  The only information they receive is the error message itself.  If worded correctly the error message can be enough (e.g. if it includes "Invalid" in the text) but this requires component users to be aware of this fact and actively account for it.

Without a well-worded error message, this results in a different experience for SR users than visual users -- the input is flagged as invalid and displayed prominently to visual users as invalid, but screen reader users get no native notification at all when they focus the input.

I think @zelliott might have a good point here that we may wish to do `(empty && required)` rather than simply `empty` for our `null` state.